iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -139,8 +144,328 @@ static Value flatten(PatternRewriter &rewriter, Location loc, Value vector) {
   return rewriter.create<vector::ShapeCastOp>(loc, dstType, vector);
 }
 
-/// Converts matrix-times-matrix-transposed vector.contracts with
-/// lhs and rhs inputs defined by arith.extsi promoting from i8 to i32,
+// Asserts that i is a power of two, and returns its log2.
+// Note: the llvm helpers used internally operate on uint32, but we keep that
+// an internal detail as the surrounding code here is all operating on signed
+// integers and mixing signed and unsigned would be error-prone.
+int8_t exactLog2(int32_t i) {
+  assert(i > 0);
+  uint32_t u = i;
+  assert(llvm::isPowerOf2_32(u));
+  return llvm::countTrailingZeros(u);
+}
+
+// Helper to handle powers of two size computations without the overhead
+// of runtime divisions. Divisions remain very expensive compared to most other
+// instructions.  Divisions are of course cheap when the divisor is
+// a constant, but a typical use case for us is
+//
+//    lhsBitWidth / kernel.registerBitWidth
+//
+// kernel.registerBitWidth is *initialized* from a literal value (say 128) but
+// it would be cumbersome to have to preserve its constant-expression status
+// throughout.
+class PowerOfTwo {
+ private:
+  int8_t exponent = 0;
+
+ public:
+  PowerOfTwo() {}
+  explicit PowerOfTwo(int32_t i) : exponent(exactLog2(i)) {}
+  int getExponent() const { return exponent; }
+  int val() const {
+    assert(exponent < 8 * sizeof(int) - 1);
+    return 1 << exponent;
+  }
+};
+
+// Returns i/p, asserting that p divides i. Requires i >= 0.
+// Fast: bit shift, not actual div.
+int fastExactDiv(int i, PowerOfTwo p) {
```

Comment:
Here (unlike the `operator*`) it's really different from the underlying integer `operator/`, in that it's more restricted and less slow.  If it were `operator/`, future people using it might feel compelled to write comments noting that this isn't an ordinary integer division, so that future-future people don't inadvertently rewrite that as just one. So the explicit function name helps here, by encoding this useful information in code rather than in comments.